netdata-installer.sh

```diff
@@ -1022,6 +1027,174 @@ install_go() {
 
 install_go
 
+function get_kernel_version() {
+  r="$(uname -r | cut -f 1 -d '-')"
+
+  read -r -a p <<< "$(echo "${r}" | tr '.' ' ')"
+
+  printf "%03d%03d%03d" "${p[0]}" "${p[1]}" "${p[2]}"
+}
+
+detect_libc() {
+  libc=
+  if ldd --version 2>&1 | grep -q -i glibc; then
+    echo >&2 " Detected GLIBC"
+    libc="glibc"
+  elif ldd --version 2>&1 | grep -q -i 'gnu libc'; then
+    echo >&2 " Detected GLIBC"
+    libc="glibc"
+  elif ldd --version 2>&1 | grep -q -i musl; then
+    echo >&2 " Detected musl"
+    libc="musl"
+  else
+    echo >&2 " ERROR: Cannot detect a supported libc on your system!"
+    return 1
+  fi
+  echo "${libc}"
+  return 0
+}
+
+get_compatible_kernel_for_ebpf() {
+  kver="${1}"
+
+  # XXX: Logic taken from Slack discussion in #ebpf
+  # everything that has a version <= 4.14 can use the code built to 4.14
+  # Continue the logic, everything that has a version <= 4.19 and >= 4.15 can use the code built to 4.19
+  # Finally,  everybody that is using 5.X can use what is compiled with 5.4
+
+  kpkg=
+
+  if [ "${kver}" -ge 005000000 ]; then
+    echo >&2 " Using eBPF Kernel Package built against Linux 5.4"
+    kpkg="5_4"
+  elif [ "${kver}" -ge 004015000 ] && [ "${kver}" -le 004020017 ]; then
+    echo >&2 " Using eBPF Kernel Package built against Linux 4.19"
+    kpkg="4_19"
+  elif [ "${kver}" -le 004014999 ]; then
+    echo >&2 " Using eBPF Kernel Package built against Linux 4.14"
+    kpkg="4_14"
+  else
+    echo >&2 " ERROR: Cannot detect a supported kernel on your system!"
+    return 1
+  fi
+
+  echo "${kpkg}"
+  return 0
+}
```

Comment:
The more I think a bout it, the more I think we should probably be installing the code for all the kernel variants we support and selecting at runtime which to load. It's not unusual for people to upgrade kernels on systems, and the current arrangement will result in a kenrel upgrade breaking the eBPF collectors until a user re-installs Netdata.

Reply:
As stated in #ebpf and #core-team and our SRE Sync today; I agree. We have a few different approaches we can take here:

- Install all variants and put the right files in place depending on the current Kernel and libc.
- Teach the agent how to load the right variants itself (_C level changes?_)
- Install all variants and put the right files in place depending on the current Kernel and libc and teach the agent how to call out to a shell script that "redoes this" if/when the Kernel changes.
- A variant of the previous but download on demand rather than install all variants.

I'm in favour of the last two personally and I'm indifferent on "install all variants" or "download on demand". I can see both approaches working equally as well and @thiagoftsm and I have already talked about the possibility of "building on demand" too remotely.

Lets discuss further after the `v1.20.0` release in #ebpf and/or a quick meeting and draft up a feature request / issue.

Reply:
I prefer "A variant of the previous but download on demand rather than install all variants.", considering that you liked it, let us move in front with it. :)

Reply:
Sounds fine to me. @thiagoftsm Do you mind creating an issue around this? Its super unclear ot me when / how we'll prioritise it; but we have a list of future enhancements we want to make too including full e2e VM testing so we're more confident what we're shipping will work for our users :) -- I'd bucket this issue/feature along with that too somewhat.

Reply:
All right, I will create in few minutes!

Reply:
Link it here when done and mark this thread as "resolved" :)

Reply:
Created => https://github.com/netdata/netdata/issues/8118

Reply:
Thank you!